cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp

```diff
@@ -0,0 +1,115 @@
+        if constexpr (std::is_same_v<Method, kmeans_init::method::plus_plus_dense>) {
+            throw unimplemented(
+                msg::kmeans_init_plus_plus_dense_method_is_not_implemented_for_gpu());
+        }
+
+        if constexpr (std::is_same_v<Method, kmeans_init::method::parallel_plus_dense>) {
+            throw unimplemented(
+                msg::kmeans_init_parallel_plus_dense_method_is_not_implemented_for_gpu());
+        }
+        return sycl::event();
+    }
+};
+
+template <typename Float>
+struct kmeans_init_kernel<Float, kmeans_init::method::dense> {
+    static sycl::event compute_initial_centroids(sycl::queue& queue,
+                                                 const pr::ndview<Float, 2>& data,
+                                                 pr::ndview<Float, 2>& centroids) {
+        ONEDAL_ASSERT(data.get_dimension(1) == centroids.get_dimension(1));
+        ONEDAL_ASSERT(data.get_dimension(0) >= centroids.get_dimension(0));
+        const std::int64_t cluster_count = centroids.get_dimension(0);
+        const std::int64_t column_count = centroids.get_dimension(1);
+        dal::detail::check_mul_overflow(cluster_count, column_count);
+        const auto data_ptr = data.get_data();
+        auto centroids_ptr = centroids.get_mutable_data();
+        return bk::copy(queue, centroids_ptr, data_ptr, cluster_count * column_count);
+    }
+};
+
+template <typename Float>
+struct kmeans_init_kernel<Float, kmeans_init::method::random_dense> {
+    static sycl::event compute_initial_centroids(sycl::queue& queue,
+                                                 const pr::ndview<Float, 2>& data,
+                                                 pr::ndview<Float, 2>& centroids) {
+        ONEDAL_ASSERT(data.get_dimension(1) == centroids.get_dimension(1));
+        ONEDAL_ASSERT(data.get_dimension(0) >= centroids.get_dimension(0));
+        const std::uint64_t row_count =
+            dal::detail::integral_cast<std::uint64_t>(data.get_dimension(0));
+        const std::uint64_t cluster_count =
+            dal::detail::integral_cast<std::uint64_t>(centroids.get_dimension(0));
+        const std::int64_t column_count = centroids.get_dimension(1);
+        const auto data_ptr = data.get_data();
+        auto centroids_ptr = centroids.get_mutable_data();
+        dal::detail::check_mul_overflow(
+            cluster_count,
+            dal::detail::integral_cast<std::uint64_t>(sizeof(std::int32_t)));
+
+        auto indices = pr::ndarray<size_t, 1>::empty(queue, cluster_count);
+        partial_fisher_yates_shuffle(indices, row_count);
+        auto indices_ptr = indices.get_data();
+
+        const std::int64_t required_local_size = bk::device_max_wg_size(queue);
+        const std::int64_t local_size = std::min(bk::down_pow2(column_count), required_local_size);
+
+        auto gather_event = queue.submit([&](sycl::handler& cgh) {
+            sycl::stream out(1024, 256, cgh);
```

Comment:
Remove?

Reply:
Removed. And fixed a hidden bug in data transfer.